### PR TITLE
Add useRef hook

### PR DESCRIPTION
**Pull request by mcordova47**
Follow-up to #8: adds a `useRef` hook, which uses `useState` to return an element along with a setter for that element, as suggested in [this comment](https://github.com/collegevine/purescript-elmish-hooks/pull/8#discussion_r777281810).

Apr-12-2022 23-01-01

### CODE REVIEW

File: examples/src/Examples/UseMouseMove.purs [hunk 1]
```diff
@@ -23,7 +23,7 @@ view =
     , H.code "" "mkHook"
     , H.text " to make a custom hook."
     ]
-  , H.div_ "w-100 py-6 rounded bg-light border position-relative"
+  , H.div_ "w-100 py-6 rounded bg-light border position-relative overflow-hidden"
```
Comment (mcordova47) [inline examples/src/Examples/UseMouseMove.purs:26]: Just a slight, drive-by improvement to an example.

File: src/Elmish/Hooks/UseRef.purs [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+-- | `elmish-html` expects `ref` to be a string.
+-- |
+-- | ```purs
+-- | view :: ReactElement
+-- | view = withHooks do
+-- |   inputEl /\ setInputEl <- useRef Nothing
+-- |   let onButtonClick = traverse_ focus inputEl
+-- |   pure $
+-- |     H.fragment
+-- |     [ H.input_ "form-control" { ref: unsafeCoerce setInputEl, defaultValue: "" }
```
Comment (fsoikin) [inline src/Elmish/Hooks/UseRef.purs:33]: How about giving the `ref` prop on `elmish-html` the right type so you don't have to use `unsafeCoerce`?

**Review by fsoikin — APPROVED** [review]
Looks good.
How about reexporting it from the root module?

File: src/Elmish/Hooks/UseRef.purs (hunk 2 above)
Comment (mcordova47) [reply src/Elmish/Hooks/UseRef.purs:33]: Yeah, but what do we want to say is the right type? `String` is allowed, but I think deprecated, so probably the worst choice. But then there’s the `{ current :: a }` thing which is used by `useRef` and `createRef`. Should we say that this is the officially supported way?

Comment (mcordova47) [reply src/Elmish/Hooks/UseRef.purs:33]: Oh, maybe it could be an opaque type and `elmish-html` could export a `functionRef :: EffectFn1 HTMLElement Unit -> Ref`?

Comment (fsoikin) [reply src/Elmish/Hooks/UseRef.purs:33]: Well, first, I do think the function should be the officially supported way, because it's, well, _functional_. The container object requires an effect for initialization, so it will always be awkward to use.

But yes, opaque type is also a fine idea.

Comment (mcordova47) [reply src/Elmish/Hooks/UseRef.purs:33]: Okay, to keep it simple, then, I’ll start with `EffectFn1 HTMLElement Unit` being the supported type 👍🏻

Comment (CharlesTaylor-CV) [reply src/Elmish/Hooks/UseRef.purs:33]: I looked at the docs on this, apparently `createRef` & `useRef` are newer. The callback ref approach has the benefit of supporting more versions of React. 

Screen Shot 2022-04-13 at 4 25 26 PM
From: https://reactjs.org/docs/refs-and-the-dom.html#dont-overuse-refs

Comment (CharlesTaylor-CV) [reply src/Elmish/Hooks/UseRef.purs:33]: String refs are definitely deprecated, like Marc said.
https://reactjs.org/docs/refs-and-the-dom.html#legacy-api-string-refs

Comment (CharlesTaylor-CV) [reply src/Elmish/Hooks/UseRef.purs:33]: > Oh, maybe it could be an opaque type and `elmish-html` could export a `functionRef :: EffectFn1 HTMLElement Unit -> Ref`? Then we could use it here and return a `Hook UseRef (Maybe HTMLElement /\ Elmish.HTML.Ref)`.

It sounds like this approach would preserve the flexibility that React refs have in the underlying framework.

Like instead of picking one opinionated way of creating refs, I vote having methods like this one to create opaque refs. 
It's less churn the next time React deprecates an api or adds a new method to create refs. There's about 4 ways to make one right now.

Comment (mcordova47) [reply src/Elmish/Hooks/UseRef.purs:33]: Alright, yeah, that makes sense. For now I’ll make an opaque type and one way to create that type (`Elmish.HTML.callbackRef`).

Comment (fsoikin) [reply src/Elmish/Hooks/UseRef.purs:33]: In that case, I think the opaque type should go into `elmish`, not `elmish-html`. And since we already have a type named `Ref` (with completely different semantics), I think we need to rename it. I propose `Opaque`. I think it's ok if it's not immediately intuitive: since the concept behind it is also not readily obvious, it would take some studying to understand it anyway.

And the new type should go, I think, under `Elmish.React`. But still reexported from the root.

Comment (mcordova47) [reply src/Elmish/Hooks/UseRef.purs:33]: Now I’m wondering if it can even work. Think about the way you deal with each type: For `callbackRef`, you have a separate value in the state that gets set. For `createRef`, the value is the same as the one you pass to the `ref` prop. So that means you need a `readRef :: Ref -> Maybe HTMLElement`, right? But that doesn’t make sense for `Ref`s created via `callbackRef`. I guess it could still return `Nothing`, but it just seems weird that a function could get passed to `readRef`. I’m leaning towards just doing `EffectFn1 HTMLElement Unit` for now and punting on deciding what the type should be as React evolves.

Comment (fsoikin) [reply src/Elmish/Hooks/UseRef.purs:33]: For `createRef` there would just be an intermediate type.

```
createRef :: Effect CellRef
readRef :: CellRef -> Maybe Element

cellRef :: CellRef -> Ref
callbackRef :: Dispatch Element -> Ref
```

But yes, this kind of awkwardness is exactly why I didn't want to do the opaque type in the first place.

Comment (fsoikin) [reply src/Elmish/Hooks/UseRef.purs:33]: Oh, and by the way: the type will have to be generic, i.e. `Ref a`, because it's not only `HTMLElement`s that can be accessed through it. Just yesterday, Charles used it to call methods on the slider component.

Comment (mcordova47) [reply src/Elmish/Hooks/UseRef.purs:33]: Ohh that is a good idea. But yeah, I think I’d rather defer for now and make a slight improvement rather than figure out what the best way to represent all this is.

Comment (fsoikin) [reply src/Elmish/Hooks/UseRef.purs:33]: And the genericity requirement applies to the naked function approach too

Comment (mcordova47) [reply src/Elmish/Hooks/UseRef.purs:33]: Oh right, well then maybe I should just make it `Foreign` like our event handler types?

Comment (fsoikin) [reply src/Elmish/Hooks/UseRef.purs:33]: Yeah, I guess that makes sense for now

Comment (mcordova47) [reply src/Elmish/Hooks/UseRef.purs:33]: But then we’re right back to `unsafeCoerce`-ing 🤔 What’s your thinking on making it generic? That’s very similar to `unsafeCoerce` in practice, because the caller can make it any type they want.

Comment (fsoikin) [reply src/Elmish/Hooks/UseRef.purs:33]: Well, no. If it's generic, then it's the props of the component that get to choose the type. So, for example:

```
type OptProps_input =
  { ...
    ...
    ref :: EffectFn1 HTMLInputElement Unit
    ...
  }

type OptProps_div =
  { ...
    ...
    ref :: EffectFn1 HTMLDivElement Unit
    ...
  }
```

And any custom component would just make their own opaque type for this:

```
foreign import data MyComponent

type SliderProps = 
  { ...
    ref :: EffectFn1 MyComponent Unit
    ...
  }

foreign import myComponentMethod :: MyComponent -> Effect Unit
```

Comment (mcordova47) [reply src/Elmish/Hooks/UseRef.purs:33]: Ah, I see what you’re saying. So `useRef` is generic. That makes sense — I’ll do that 👍🏻

Comment (mcordova47) [reply src/Elmish/Hooks/UseRef.purs:33]: Although it makes it very difficult to work with the code-gen in `elmish-html`. I’ll think about it some more.

Comment (mcordova47) [reply src/Elmish/Hooks/UseRef.purs:33]: If everything is named sanely, we can just to `Web.HTML.HTML{Capitalizedtag}Element`

**Review by CharlesTaylor-CV — APPROVED** [review]

File: src/Elmish/Hooks/UseRef.purs [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+-- | view = withHooks do
+-- |   inputEl /\ setInputEl <- useRef Nothing
+-- |   let onButtonClick = traverse_ (focus <<< HTMLInputElement.toHTMLElement) inputEl
+-- |   pure $
+-- |     H.fragment
+-- |     [ H.input_ "form-control" { ref: setInputEl, defaultValue: "" }
+-- |     , H.button_ "btn btn-primary" { onClick: onButtonClick } "Focus the input"
+-- |     ]
+-- | ```
+useRef :: forall el. Hook (UseRef el) (Maybe el /\ (EffectFn1 el Unit))
```
Comment (fsoikin) [inline src/Elmish/Hooks/UseRef.purs:33]: I'm not sure if this is for aesthetic reasons, but parent are not required:

```suggestion
useRef :: forall el. Hook (UseRef el) (Maybe el /\ EffectFn1 el Unit)
```

**Review by fsoikin — APPROVED** [review]
Awesome! 👍

**Comment by mcordova47** [comment]
Updated to use the new opaque `ReactRef` type. Have to merge collegevine/purescript-elmish#56 and collegevine/purescript-elmish-html#31 and update versions here before merging this.
